src/test/java/io/jenkins/pluginhealth/scoring/probes/ProbeEngineTest.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2022 Jenkins Infra
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.jenkins.pluginhealth.scoring.probes;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.atMostOnce;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.stream.Stream;
+
+import io.jenkins.pluginhealth.scoring.model.Plugin;
+import io.jenkins.pluginhealth.scoring.model.ProbeResult;
+import io.jenkins.pluginhealth.scoring.service.PluginService;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class ProbeEngineTest {
+
+    @Mock
+    private PluginService pluginService;
+
+    @Test
+    public void shouldBeAbleToRunSimpleProbe() {
+        final Plugin plugin = mock(Plugin.class);
+        final Probe probe = mock(Probe.class);
+        final ProbeEngine probeEngine = new ProbeEngine(List.of(probe), pluginService);
+        final ProbeResult expectedResult = ProbeResult.success("foo", "bar");
+
+        when(probe.doApply(plugin)).thenReturn(expectedResult);
+        when(pluginService.streamAll()).thenReturn(Stream.of(plugin));
+        probeEngine.run();
+
+        verify(probe).doApply(plugin);
+        verify(plugin, atMostOnce()).addDetails(expectedResult);
+        verify(pluginService, atMostOnce()).saveOrUpdate(plugin);
```

Comment:
The verify method is used to check how many times a method of a given mock object has been invoked.
So, in this case, we are cross-checking something like: "Hey, when we called probeEngine.run(); we had to call the methods `doApply`, addDetails` and `saveOrUpdate` at most once (once, in case of "doApply") in the subsequent calls." Right?

In this way we're testing what happens when we call the `probeEngine.run()` method, we cannot do end-to-end testing so that's why we're relying on the amount of times a method is called to make sure everything works as expected.

Reply:
exactly my thinking. I don't want to code a dummy probe and see that it provided a result. What I want is to see that the probe engine called the method `doApply` on the probe with the provided plugin, that it added the `ProbeResult` to the plugin and that we saved the plugin.